content/doc/developer/testing/index.adoc

```diff
@@ -110,7 +110,28 @@ Most Java IDEs should be able to run JUnit tests and report on the results.
 == What to Test
 Now that we can write a basic test, we should discuss what you should be testing…
 
-TODO: Unit testing of your code, as much as possible. JenkinsRule testing: creating jobs that use your build steps, running them, asserting on the output
+The test pyramid is a concept used in software development to describe the ideal distribution of different types of automated tests that make up a test suite.
+The idea was introduced by Mike Cohn and is visualized as a pyramid to help illustrate the optimal proportion of each type of test.
+Here's a breakdown of the three main layers from the bottom to the top of the pyramid:
+
+1. *Unit Tests*: At the base of the pyramid, unit tests are the most numerous.
+These tests are focused on checking the smallest parts of an application, such as individual functions or methods.
+Unit tests are quick to execute and help ensure that each component of the software works as expected in isolation.
+
+2. *Integration Tests*: The middle layer of the pyramid consists of integration tests.
+These are fewer in number compared to unit tests.
+Integration tests verify that different parts of the system work together as intended.
+For Jenkins, this makes use of the `JenkinsRule` to test the integration of the plugin into Jenkins.
+This can be a test for a new build step and testing this with a test pipeline.
+
+3. *End-to-End Tests (E2E) and UI Tests*: At the top of the pyramid, end-to-end tests are the fewest in number but typically the most complex.
+These tests simulate real user scenarios from start to finish.
+They interact with the application as a user would, testing the complete flow of the system.
+For Jenkins, this makes use of the `JenkinsRule` or `JenkinsSessionRule` in combination with the `createWebClient()`-call which is used to really creates a call to endpoint to assert something on the result.
+
+Purpose of the Test Pyramid: The pyramid aims to encourage developers to write a larger number of lower-level tests (unit tests) and fewer high-level tests (E2E tests). This distribution is recommended because lower-level tests tend to be faster, cheaper to automate, and more reliable, whereas higher-level tests are slower, more expensive, and can be flakier.
```

Comment:
We've preferred to use sentence per line for easier formatting.  Optional change, not required.
```suggestion
Purpose of the Test Pyramid: The pyramid aims to encourage developers to write a larger number of lower-level tests (unit tests) and fewer high-level tests (E2E tests).
This distribution is recommended because lower-level tests tend to be faster, cheaper to automate, and more reliable, whereas higher-level tests are slower, more expensive, and can be flakier.
```